Update luci.py device tracker to work with OpenWRT 18.06

fbradyirl opened the pull request:
## Description:
Updating to use new APIs in OpenWRT 18.06 (which was released yesterday). 

The old `net.arptable` method is no longer available so device tracking will be broken in HA for users who upgrade their router to the latest.

**Related issue (if applicable):** fixes #8508

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

fbradyirl commented:
cc previous editors to this device tracker @pvizeli @fingon

homeassistant commented:
Hello @fbradyirl,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`af9871e9b695d16d8edeeed5d313ee900b15ee14`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;15792&#x2F;commits&#x2F;af9871e9b695d16d8edeeed5d313ee900b15ee14) - This commit has something that looks like an email address (fibrady@cisco.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @fbradyirl,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`e2faa4a7393dac9a16c544cc08990c2ec60ea113`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;15792&#x2F;commits&#x2F;e2faa4a7393dac9a16c544cc08990c2ec60ea113) - No email found attached to the commit.
- [`af9871e9b695d16d8edeeed5d313ee900b15ee14`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;15792&#x2F;commits&#x2F;af9871e9b695d16d8edeeed5d313ee900b15ee14) - This commit has something that looks like an email address (fibrady@cisco.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

fbradyirl committed:
Update luci.py

    Updating to use new APIs in OpenWRT 18.06. The old net.arptable method is no longer available.

fbradyirl committed:
Fix hound issue

fbradyirl committed:
Fix hound issue

fbradyirl committed:
Fix hound issue

fbradyirl committed:
Remove luci from coveragerc file

fbradyirl committed:
Adding json fixtures for tests

fbradyirl committed:
Fix issues

lkollar reviewed (commented):
Would it make sense to create two subclasses which inherit from `LuciDeviceScanner` and implement the pre-18.06 and post-18.06 logic by overriding methods in the base class? `get_scanner` could then invoke a factory method on `LuciDeviceScanner` to return the right implementation.

fbradyirl commented:
@lkollar Yes I like that idea. I'll try and get time to implement that over the coming days. Thanks for looking!

MartinHjelmare reviewed (changes requested):
Please break out device specific code, requests and data parsing, into a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

We shouldn't make any changes to this module until that is done.

fabaff commented:
Can somebody please comment on https://github.com/home-assistant/home-assistant.io/pull/6050? UBUS vs. RPC for retrieving details. Thanks

rytilahti commented:
Just a side-note, I was working to externalize this but found out that there's a push-based solution which is much nicer way for device tracking (https://github.com/openwrt/packages/pull/5701). As part of that effort I created https://github.com/rytilahti/python-ubus for communicating with the ubus-rpc, which may be interesting for the externalized luci tracker lib.

neographikal commented:
Noticed this bug this weekend as well, present on HA 0.81.4 and Openwrt 18.06.1 as well.

marine1988 commented:
i have last  version of openwrt and im having this issues my device iss allways home :( any fix

marine1988 commented:
any news?

disrupted commented:
@fbradyirl why did you close it?

fbradyirl commented:
@disrupted I closed it because I don't have the time at present to implement the work as a pip module. I can reopen if I ever get the chance, but right now, it is not happening, so best to close the PR.

marine1988 commented:
what is the PR?

Finbarr Brady <notifications@github.com> escreveu no dia terça, 20/11/2018
à(s) 12:17:

> @disrupted <https://github.com/disrupted> I closed it because I don't
> have the time at present to implement the work as a pip module. I can
> reopen if I ever get the chance, but right now, it is not happening, so
> best to close the PR.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/15792#issuecomment-440253224>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGOMpByrqp6NlgwlYnJr5hSeP6K6gzg0ks5uw_K9gaJpZM4Vr-nK>
> .
>

VergLsm commented:
@fbradyirl We waiting for you reopen it;

reaper7 commented:
I spent two days looking a solution for luci device tracker (after updating my router openwrt to 18.06.01),
and finally I found @fbradyirl PR

It just works! Please implement this

flipreverse commented:
Please fix this issue.
Why can't you just merge the PR?

fbradyirl commented:
@flipreverse Converting my branch into a pip module as that's the way external APIs are to be interacted with from HA.

I've set up a template project for the pip module here:
https://github.com/fbradyirl/openwrt-luci-rpc/

I've added you as a collab, so if you can help that would be good. I am just busy at present but might eventually get to it.

fbradyirl commented:
> Why should it be a pip module?
@flipreverse see the above PR feedback...

> Please break out device specific code, requests and data parsing, into a standalone library published on pypi.
> 
> https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services
> 
> We shouldn't make any changes to this module until that is done.

fbradyirl commented:
Yes as I said earlier, my device doesn’t have wifi so any tracker based off that won’t sork for everyone. Thanks anyway!

reaper7 commented:
@fbradyirl - true
we need complex solution for non WiFi devices too...
https://github.com/mueslo/openwrt_hass_devicetracker/issues/15

fbradyirl commented:
Agreed. Falling back to the arp table seems to be the next best thing without WiFi.

DeanoXX commented:
Just another vote/request for a 18.06 compliant tracker.

fbradyirl commented:
Opened new PR with the fix.

flipreverse commented:
Thank you @fbradyirl  for fixing this!

MartinHjelmare commented:
I'm going to lock this thread now. The new PR is merged. If there are bugs please open new issues.
